### PR TITLE
[SILGen] Create SIL variable declaration scopes for trivial values.

### PR DESCRIPTION
**Pull request by atrick**
[SILGen] emit local variable scopes for trivial values.

Required for SIL analysis of local variable scopes. For example, to track lifetime dependence on a variable.

Example:

.swift:

    func foo(i: Int) {
      let v = i

.sil:

```
    %1 = move_value [var_decl] %0 : $Int // starts the variable scope
    extend_lifetime %1 : $Int            // ends the variable scope
```

### CODE REVIEW

**Review by nate-chandler — APPROVED** [review]
LGTM

File: lib/SIL/IR/ValueOwnership.cpp [hunk 1]
```diff
@@ -212,6 +210,12 @@ CONSTANT_OR_NONE_OWNERSHIP_INST(Guaranteed, OpenExistentialBoxValue)
 // value).
 CONSTANT_OR_NONE_OWNERSHIP_INST(Owned, MarkUninitialized)
 
+// In raw SIL, a MoveValue delimits the scope of trivial variables.
+CONSTANT_OR_NONE_OWNERSHIP_INST(Owned, MoveValue)
+
+// In raw SIL, a BeginBorrow delimits the scope of trivial variables.
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SIL/IR/ValueOwnership.cpp:216]: Maybe we won't need to support `begin_borrow : $Trivial`.
Also, it might be better to introduce a new opcode for trivial `move_value`. I originally wanted as much consistency as possible between trivial/nontrivial vardecls in diagnostics. If we want perfect consistency though, the variable scope would use separate scope markers: `%token = begin_vardecl` -> `end_vardecl %token` instead of flags. `begin_vardecl` would only produce a token, so we don't need to add use-def handling, and those scoped can be dropped after diagnostics. But I don't know if it's realistic at this point to change all code that relies on move_value/begin_borrow.

[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SIL/IR/ValueOwnership.cpp:216]: Maybe it would be best to postpone permitting `begin_borrow %t : $Trivial` until it actually gets used.

[the remark this answers is not in the document]
Comment (atrick) [reply lib/SIL/IR/ValueOwnership.cpp]: I was convinced that we needed `begin_borrow` until after I rewrote `getValueForLexicalLifetimeBinding`

File: lib/SILGen/SILGenDecl.cpp [hunk 2]
```diff
@@ -809,50 +809,26 @@ class LetValueInitialization : public Initialization {
...
-        value = SGF.B.createMoveValue(PrologueLoc, value, IsLexical);
-        return SGF.B.createMarkUnresolvedNonCopyableValueInst(
-            PrologueLoc, value,
-            MarkUnresolvedNonCopyableValueInst::CheckKind::
-                ConsumableAndAssignable);
-      }
-
-      // If we have a no implicit copy trivial type, wrap it in the move only
-      // wrapper and mark it as needing checking by the move checker.
-      if (vd->isNoImplicitCopy() && value->getType().isTrivial(SGF.F)) {
```
Comment (atrick) [inline lib/SILGen/SILGenDecl.cpp:831]: @nate-chandler before my rewrite of `getValueForLexicalLifetimeBinding`, noImplicitCopy variables were considered to have lexical lifetimes. Was that intentional? I thought a goal of noImplicitCopy was to be fully optimized, thus eagerly consumed...

Old code:
```
    if (vd->isNoImplicitCopy()) {
      value = SGF.B.createMoveValue(PrologueLoc, value, IsLexical,
                                    DoesNotHavePointerEscape, IsFromVarDecl);
```

Comment (nate-chandler) [reply lib/SILGen/SILGenDecl.cpp:831]: I don't know.  Looking at the blame, that behavior was added as
```
+      // Otherwise, we have a no implicit copy trivial type, so wrap it in the
+      // move only wrapper and mark it as needing checking by the move cherk.
+      value =
+          SGF.B.createOwnedCopyableToMoveOnlyWrapperValue(PrologueLoc, value);
+      value = SGF.B.createMoveValue(PrologueLoc, value, /*isLexical*/ true);
```
along with the entirety of the original method `getValueForLexicalLifetimeBinding` that seems to have enjoyed mostly NFC changes since then in

```
commit a32c5b3277dbffa6946d117d964e3764d285915e
Date:   Fri Jul 22 11:49:37 2022 -0700

    [silgen] Refactor out a helper method from LetValueInitialization::bindValue.

    Specifically, I refactored out the logic around setting up the value in the case
    of having lexical lifetimes/no implicit copy into the method
    LetValueInitialization::getValueForLexicalLifetimeBinding.

    This logic was taking over bindValue and was making the logic hard to follow.
    Additionally this allows me to take advantage of return value/early exits to
    make the actual logic in getValueForLexicalLifetimeBinding easier to reason
    about.

    NFC.
```

Comment (atrick) [reply lib/SILGen/SILGenDecl.cpp:831]: > I don't know. Looking at the blame, that behavior was added as

@nate-chandler You don't need to justify the code. I was just wondering if I'm misremembering some design change. As it turns out, yes I was misremembering and I need to revert this change.

Originally no-implicit-copy was going to be eager-consume, but later we changed  course. The only copies that eager-move would be able to eliminate would be explicit copies, and eliminating those is a non-goal. Furthermore, people expect these variables to be destroyed at an explicit `consume` operator, and not earlier than that relative to deinit barriers. To get eager-consume behavior, the programmer simply needs to use an explicit `consume` operator rather that an explicit `copy`.

That's confusing, but basically no-implicit-copy turns on explicit lifetime management. It does *not* turn on automatic aggressive lifetime optimization.

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
source compate penny-bot_510 also fails on main

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test linux platform
